Fix CMSMenuItem CSS references for namespaced controllers

nedmas opened the pull request:
The controller class name including namespace is used to create CSS class names for `CMSMenuItem` (specifically for the icon).

Currently this is done without escaping the class name resulting in CSS classes with backslashes (`\`) which are invalid characters for CSS class names.

This PR simply uses the existing `Convert::raw2htmlname()` method to escape the class name before use in `CMSMenuItem`.

nedmas commented:
@tractorcow any chance this can be included in 3.1.14 please?

dhensby commented:
Please can you tell me what the before and after class names are?

nedmas commented:
Sure, so `Convert::raw2htmlname()` simply strips out the backslashes (`\`).

Example before:
`Example\Namespace\Controllers\ExampleAdmin`

Example after:
`ExampleNamespaceControllersExampleAdmin`

nedmas commented:
I've updated the PR to include the same stripping of backslashes (`\`) from the controller name when used to reference the ID and CSS class names when updating the menu. Aside from this I can find no other uses of the class name that need to be updated.

dhensby commented:
OK, I think that `\` need to be replaced with some character, otherwise a controller named `MyController` would conflict with a controller named `My\Controller`.

nedmas commented:
Agreed, although I think a conflict is unlikely we should code defensively. What character is best to use? I would suggest underscore (`_`).

dhensby commented:
I think underscores do make sense... but then PSR0 named classes could ALSO conflict.. gosh.

I reckon we go with an `_` and that'll be close enough

nedmas commented:
Good point, maybe we should use dash (`-`) instead? Might as well avoid all known conflicts. Can you think of any potential conflicts using dash (`-`)?

dhensby commented:
That seems sensible :) can't think of anything immediately

nedmas commented:
I've updated the PR to replace backslashes (`\`) with dashes (`-`). I've extracted it into a helper function called `Convert::fqcn2htmlname()` as it was needed in more than one class. The helper function replaces all backslashes with dashes then runs it through `Convert::raw2htmlname()` for good measure. I've added the same regex used in `Convert::raw2htmlname()` after replacing the backslashes with dashes in the javascript as well.

Is there anything else that I should include or anything I've missed?

dhensby commented:
If you're going to add new features then it can't make a patch release (so this PR will need to go against 3 (or 3.2)).

If you want it in a 3.1 release, keep this to the minimum required to fix the bug to get it into 3.1.x release

dhensby commented on core/Convert.php:
```diff
@@ -353,4 +355,19 @@ public static function raw2url($title) {
 		return preg_replace('~\R~u', $nl, $data);
 	}
+
+	/*
```
PHP doc block should start `/**`

dhensby commented on core/Convert.php:
```diff
@@ -353,4 +355,19 @@ public static function raw2url($title) {
+	 * @return string
+	 */
+	public function fqcn2htmlname($val)
+	{
```
This is pedantic, but the coding conventions require the `{` to be on the same line as the `)`

nedmas commented:
Ok I'll change it to exclude new features and add them in against 3 or 3.2 once this is merged.

dhensby commented:
Thanks :)

nedmas commented:
Thanks
